File: packages/native-android/app/src/main/java/com/example/wearos/presentation/SensorModules.kt
package com.example.wearos.presentation

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import com.wrst.runtime.WrstNativeModules
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

// Example "custom native modules": scalar sensors that aren't engine built-ins.
// Each is a streaming module - JS subscribes via subscribeNativeModule(name), the
// module pushes { value, timestamp } per sample through WrstNativeModules.emit().
// These are SensorManager sensors, so one generic helper covers them.
//
// Kept cross-platform (also in iOS SensorModules.swift): barometer needs no
// permission; stepCount needs ACTIVITY_RECOGNITION ("activity"), requested from
// JS. (Android can also do heartRate/ambientLight, but those aren't symmetric
// with iOS, so the example omits them.)
object SensorModules {
    private lateinit var sensorManager: SensorManager
    private val listeners = ConcurrentHashMap<String, SensorEventListener>()

    fun register(context: Context) {
        sensorManager =
            context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        registerSensor("barometer", Sensor.TYPE_PRESSURE)       // hPa
        registerSensor("stepCount", Sensor.TYPE_STEP_COUNTER)   // cumulative steps
    }

    private fun registerSensor(name: String, sensorType: Int) {
        WrstNativeModules.register(name) { args ->
            val cmd = args.getOrNull(0) as? JSONObject ?: return@register null
            val callbackId = cmd.optString("callbackId")
            when (cmd.optString("action")) {
                "start" -> start(callbackId, sensorType, cmd.optLong("intervalMs", 1000L))
                "stop" -> stop(callbackId)
            }
            null
        }
    }

    private fun start(callbackId: String, sensorType: Int, intervalMs: Long) {
        if (callbackId.isEmpty()) return
        val sensor = sensorManager.getDefaultSensor(sensorType) ?: return
        stop(callbackId)
        var lastEmit = 0L
        val listener = object : SensorEventListener {
            override fun onSensorChanged(event: SensorEvent) {
                val now = System.currentTimeMillis()
                if (now - lastEmit < intervalMs) return
                lastEmit = now
                val value = event.values.getOrElse(0) { 0f }
                WrstNativeModules.emit(
                    callbackId,
                    """{"value":$value,"timestamp":$now}""",
                )
            }

            override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
        }
        listeners[callbackId] = listener
        sensorManager.registerListener(listener, sensor, (intervalMs * 1000).toInt())
    }

    private fun stop(callbackId: String) {
        listeners.remove(callbackId)?.let { sensorManager.unregisterListener(it) }
    }
}

File: packages/native-ios/AppleWatch Watch App/SensorModules.swift
import Foundation
import CoreMotion
import WrstRuntime

// Example "custom native modules" for the sensors that aren't engine built-ins.
// Each is a streaming module: JS subscribes via subscribeNativeModule(name) and
// the module pushes { value, timestamp } per sample through WrstNativeModules.emit().
//
// barometer  -> CMAltimeter (hPa),  stepCount -> CMPedometer (cumulative steps).
// Both need NSMotionUsageDescription (declared in Info.plist). heartRate and
// ambientLight are intentionally not implemented on watchOS: live heart rate
// needs an HKWorkoutSession + the HealthKit entitlement, and there's no public
// ambient-light API - so subscribing to them here simply yields no samples.
@MainActor
enum SensorModules {
    private static let altimeter = CMAltimeter()
    private static let pedometer = CMPedometer()
    private static var barometerCallback: String?
    private static var stepCallback: String?

    static func register() {
        WrstNativeModules.register("barometer") { args in
            MainActor.assumeIsolated { handleBarometer(args) }
            return nil
        }
        WrstNativeModules.register("stepCount") { args in
            MainActor.assumeIsolated { handleSteps(args) }
            return nil
        }
    }

    private static func command(_ args: [Any]) -> (action: String, callbackId: String)? {
        guard let dict = args.first as? [String: Any],
              let action = dict["action"] as? String,
              let callbackId = dict["callbackId"] as? String else { return nil }
        return (action, callbackId)
    }

    private static func handleBarometer(_ args: [Any]) {
        guard let (action, callbackId) = command(args) else { return }
        switch action {
        case "start":
            guard CMAltimeter.isRelativeAltitudeAvailable() else { return }
            barometerCallback = callbackId
            altimeter.startRelativeAltitudeUpdates(to: .main) { data, _ in
                guard let kPa = data?.pressure.doubleValue else { return }
                MainActor.assumeIsolated {
                    guard let id = barometerCallback else { return }
                    WrstNativeModules.emit(id, "{\"value\":\(kPa * 10),\"timestamp\":\(now())}")
                }
            }
        case "stop":
            if barometerCallback == callbackId {
                altimeter.stopRelativeAltitudeUpdates()
                barometerCallback = nil
            }
        default:
            break
        }
    }

    private static func handleSteps(_ args: [Any]) {
        guard let (action, callbackId) = command(args) else { return }
        switch action {
        case "start":
            guard CMPedometer.isStepCountingAvailable() else { return }
            stepCallback = callbackId
            pedometer.startUpdates(from: Date()) { data, _ in
                guard let steps = data?.numberOfSteps.doubleValue else { return }
                DispatchQueue.main.async {
                    MainActor.assumeIsolated {
                        guard let id = stepCallback else { return }
                        WrstNativeModules.emit(id, "{\"value\":\(steps),\"timestamp\":\(now())}")
                    }
                }
            }
        case "stop":
            if stepCallback == callbackId {
                pedometer.stopUpdates()
                stepCallback = nil
            }
        default:
            break
        }
    }

    private static func now() -> Double { Date().timeIntervalSince1970 * 1000 }
}
